### PR TITLE
feat(auth-public-api): Use token exchange for national registry service

### PR DESCRIPTION
**Pull request by valurefugl**
Enable token exchange in auth middleware. Use token exchange in national registry service.

## Checklist:

- [ x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/auth-nest-tools/src/lib/auth.middleware.ts [hunk 1]
```diff
@@ -18,22 +19,111 @@ interface Middleware {
...
+      return ''
+    }
+  }
+
+  private logError(errorMessage: string) {
+    if (this.logger) {
+      this.logger.error(errorMessage)
+    } else {
+      console.log(errorMessage)
+    }
```
Comment (eirikurn) [inline libs/auth-nest-tools/src/lib/auth.middleware.ts:127]: It´s possible to import a logger object from @island.is/logging directly.

File: libs/auth-nest-tools/src/lib/auth.middleware.ts [hunk 2]
```diff
@@ -18,22 +19,111 @@ interface Middleware {
...
+          client_secret: options.clientSecret,
+          subject_token: accessToken,
+          subject_token_type: 'urn:ietf:params:oauth:token-type:access_token',
+        }),
+      })
+
+      if (!response.ok) {
+        this.logError(`Token exchange failed, ${await response.text()}`)
+        return ''
+      }
```
Comment (eirikurn) [inline libs/auth-nest-tools/src/lib/auth.middleware.ts:93]: Should throw an error instead if token exchange fails?

File: apps/services/auth-public-api/src/environments/environment.ts [hunk 3]
```diff
@@ -26,6 +26,15 @@ const devConfig = {
       clientId: 'IS-DEV/GOV/10000/island-is-client',
       memberClass: XRoadMemberClass.GovernmentInstitution,
     },
+    authMiddlewareOptions: {
+      forwardUserInfo: false,
+      tokenExchangeOptions: {
+        issuer: 'https://identity-server.dev01.devland.is',
+        clientId: '@island.is/clients/national-registry',
+        clientSecret: process.env.NATIONAL_REGISTRY_IDS_CLIENT_SECRET,
+        scope: '@skra.is/individuals api_resource.scope', // TODO: remove api_resource.scope
```
Comment (eirikurn) [inline apps/services/auth-public-api/src/environments/environment.ts:35]: Why is api_resource.scope needed? What's the plan? Just curious.

File: libs/auth-nest-tools/src/lib/auth.middleware.ts (hunk 1 above)
Comment (eirikurn) [reply libs/auth-nest-tools/src/lib/auth.middleware.ts:127]: Better than having to pass in a logger and sometimes forgetting and ending with console.log

File: libs/auth-nest-tools/src/lib/auth.middleware.ts [hunk 4]
```diff
@@ -18,22 +19,111 @@ interface Middleware {
...
       context.init.headers = Object.assign({}, context.init.headers, {
         'User-Agent': this.auth.userAgent,
         'X-Real-IP': this.auth.ip,
       })
     }
   }
+
+  private async exchangeToken(
+    accessToken: string,
+    options: TokenExchangeOptions,
```
Comment (eirikurn) [inline libs/auth-nest-tools/src/lib/auth.middleware.ts:72]: Should access the options from `this` directly?

File: libs/auth-nest-tools/src/lib/auth.middleware.ts [hunk 5]
```diff
@@ -18,22 +19,111 @@ interface Middleware {
...
+  forwardUserInfo: boolean
+  tokenExchangeOptions?: TokenExchangeOptions
+}
+
+export interface TokenExchangeOptions {
+  issuer: string
+  clientId: string
+  clientSecret: string
+  scope: string
+}
```
Comment (eirikurn) [inline libs/auth-nest-tools/src/lib/auth.middleware.ts:32]: Would be nice if the token exchange could be short-circuited if the authentication already contains the necessary scopes.

File: libs/auth-nest-tools/src/lib/auth.middleware.ts (hunk 2 above)
Comment (eirikurn) [reply libs/auth-nest-tools/src/lib/auth.middleware.ts:93]: Might be good to include the requested URL from the context in the error/log.

**Review by eirikurn — CHANGES_REQUESTED** [review]
Nice to build this into the AuthMiddleware, at least for now.

File: libs/auth-nest-tools/src/lib/auth.middleware.ts (hunk 1 above)
Comment (valurefugl) [reply libs/auth-nest-tools/src/lib/auth.middleware.ts]: Throwing errors now, no logging in middleware.

File: libs/auth-nest-tools/src/lib/auth.middleware.ts (hunk 4 above)
Comment (valurefugl) [reply libs/auth-nest-tools/src/lib/auth.middleware.ts]: Fixed.

File: libs/auth-nest-tools/src/lib/auth.middleware.ts (hunk 5 above)
Comment (valurefugl) [reply libs/auth-nest-tools/src/lib/auth.middleware.ts:31]: Implemented

File: apps/services/auth-public-api/src/environments/environment.ts (hunk 3 above)
Comment (valurefugl) [reply apps/services/auth-public-api/src/environments/environment.ts:38]: This is what Þjóðskrá used in the beginning and changing it has not been a priority for them. Should be removed soon.

**Review by unnarb — APPROVED** [review]

**Comment by valurefugl** [comment]
Abandoning this, continued here, https://github.com/island-is/island.is/pull/4302

**Pull request closed without merging**